azure: add virtual machine data disk attachment resource

prfj opened the pull request:
<!--
Thank you for helping to improve Official Azure Provider!

Please read through https://git.io/fj2m9 if this is your first time opening a
Official Azure Provider pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Official Azure Provider issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #467

I have:

- [X] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

Tested locally by applying CRDs, creating a default ProviderConfig, run `make run` and apply the examples/linuxvirtualmachinedatadiskattachment.yaml file

```
$ kubectl get managed

NAME                                     READY   SYNCED   EXTERNAL-NAME   AGE
resourcegroup.azure.upbound.io/example   True    True     example         7m22s

NAME                                                   READY   SYNCED   EXTERNAL-NAME   AGE
linuxvirtualmachine.compute.azure.upbound.io/example   True    True     example         7m22s

NAME                                           READY   SYNCED   EXTERNAL-NAME   AGE
manageddisk.compute.azure.upbound.io/example   True    True     example         7m22s

NAME                                                                READY   SYNCED   EXTERNAL-NAME                                                                                                                                      AGE
virtualmachinedatadiskattachment.compute.azure.upbound.io/example   True    True     /subscriptions/XXXXXX/resourceGroups/example/providers/Microsoft.Compute/virtualMachines/example/dataDisks/example   7m22s

NAME                                                READY   SYNCED   EXTERNAL-NAME   AGE
networkinterface.network.azure.upbound.io/example   True    True     example         7m23s

NAME                                      READY   SYNCED   EXTERNAL-NAME   AGE
subnet.network.azure.upbound.io/example   True    True     example         7m23s

NAME                                              READY   SYNCED   EXTERNAL-NAME   AGE
virtualnetwork.network.azure.upbound.io/example   True    True     example         7m23s

```


Tested with [Uptest](https://github.com/upbound/provider-azure/actions/runs/6811206117)

turkenf commented:
/test-examples="examples/compute/virtualmachinedatadiskattachment.yaml"

turkenf commented on config/externalname.go:
```diff
@@ -215,8 +215,9 @@ var ExternalNameConfigs = map[string]config.ExternalName{
-	"azurerm_virtual_machine_extension": config.TemplatedStringAsIdentifier("name", "{{ .parameters.virtual_machine_id }}/extensions/{{ .external_name }}"),
+	"azurerm_managed_disk_sas_token":               config.IdentifierFromProvider,
+	"azurerm_virtual_machine_extension":            config.TemplatedStringAsIdentifier("name", "{{ .parameters.virtual_machine_id }}/extensions/{{ .external_name }}"),
+	"azurerm_virtual_machine_data_disk_attachment": config.TemplatedStringAsIdentifier("name", "{{ .parameters.virtual_machine_id }}/dataDisks/{{ .external_name }}"),
```
Since we do not have a name parameter here, I assume that this configuration will not work.

It does not seem possible to normalize the external name. Could you try the following?

```suggestion
	"azurerm_virtual_machine_data_disk_attachment": config.IdentifierFromProvider,
```

prfj replied:
i've changed to use intentifier from provider

turkenf reviewed (commented):
Thank you @prfj, I left a comment for you. 

Could you please do the manual tests of the resource and fill in the description section?

prfj commented:
@turkenf  Can you run a new uptest now ? I've added selectors/references and fixed the example file 


> Could you please do the manual tests of the resource and fill in the description section?

I've tested using the ["Manual Test" method](https://github.com/crossplane/upjet/blob/v0.10.0/docs/testing-instructions.md#manual-test) what kind of evidence you want me to put on description ?

prfj commented:
/test-examples="examples/compute/virtualmachinedatadiskattachment.yaml"

turkenf commented on config/compute/config.go:
```diff
@@ -47,6 +47,12 @@ func Configure(p *config.Provider) {
 	})
+	p.AddResourceConfigurator("azurerm_virtual_machine_data_disk_attachment", func(r *config.Resource) {
+		r.References["virtual_machine_id"] = config.Reference{
+			Type:      rconfig.APISPackagePath + "/compute/v1beta1.LinuxVirtualMachine",
```
Could you use the following configuration which is safer?

```suggestion
			TerraformName: "azurerm_linux_virtual_machine",
```

turkenf commented on examples/compute/virtualmachinedatadiskattachment.yaml:
```diff
@@ -0,0 +1,148 @@
+
```
Could you please remove this line?

turkenf commented on examples/compute/virtualmachinedatadiskattachment.yaml:
```diff
@@ -0,0 +1,148 @@
+kind: LinuxVirtualMachine
+metadata:
+  annotations:
+    meta.upbound.io/example-id: compute/v1beta1/virtualmachineextension
```
```suggestion
    meta.upbound.io/example-id: compute/v1beta1/virtualmachinedatadiskattachment
```

turkenf commented on examples/compute/virtualmachinedatadiskattachment.yaml:
```diff
@@ -0,0 +1,148 @@
+
+---
+
+apiVersion: compute.azure.upbound.io/v1beta1
+kind: VirtualMachineDataDiskAttachment
+metadata:
+  annotations:
+    meta.upbound.io/example-id: compute/v1beta1/virtualmachinedatadiskattachment
+  labels:
+    testing.upbound.io/example-name: example
+  name: example
+spec:
+  forProvider:
+    caching: ReadWrite
+    lun: 1
+    managedDiskIdSelector:
+      matchLabels:
+        testing.upbound.io/example-name: example
+    virtualMachineIdSelector:
+      matchLabels:
+        testing.upbound.io/example-name: example
```
Could you please move the `VirtualMachineDataDiskAttachment` resource example to the beginning of the YAML file?

prfj commented:
@turkenf applied sugestions please check

turkenf commented:
/test-examples="examples/compute/virtualmachinedatadiskattachment.yaml"

turkenf reviewed (approved):
Thank you for your effort @prfj, LGTM.
